tests/pyfunc/test_pyfunc_schema_enforcement.py

```diff
@@ -914,3 +914,100 @@ def test_schema_enforcement_for_list_inputs():
     pd_data = pd.DataFrame([data])
     pd_check = _enforce_schema(pd_data.to_dict(orient="list"), signature.inputs)
     pd.testing.assert_frame_equal(pd_check, pd_data)
+
+
+def test_enforce_params_schema():
+    # Correct parameters & schema
+    test_parameters = {
+        "a": "str_a",
+        "b": np.int32(1),
+        "c": True,
+        "d": 1.0,
+        "e": np.float32(0.1),
+        "f": b"byte_g",
+        "g": np.int64(100),
+        "h": np.datetime64("20230626"),
+        "i": ["a", "b", "c"],
+        "j": [True, False],
+    }
+    test_schema = ParamSchema(
+        [
+            ParamSpec("a", DataType.string, "str_a", None),
+            ParamSpec("b", DataType.integer, np.int32(1), None),
+            ParamSpec("c", DataType.boolean, True, None),
+            ParamSpec("d", DataType.double, 1.0, None),
+            ParamSpec("e", DataType.float, np.float32(0.1), None),
+            ParamSpec("f", DataType.binary, b"byte_g", None),
+            ParamSpec("g", DataType.long, np.int64(100), None),
+            ParamSpec("h", DataType.datetime, np.datetime64("20230626"), None),
+            ParamSpec("i", DataType.string, ["a", "b", "c"], (-1,)),
+            ParamSpec("j", DataType.boolean, [True, False], (-1,)),
+        ]
+    )
+    assert _enforce_params_schema(test_parameters, test_schema) == test_parameters
+
+    # Add default values if the parameter is not provided
+    test_parameters = {"a": "str_a"}
+    test_schema = ParamSchema(
+        [ParamSpec("a", DataType.string, ""), ParamSpec("b", DataType.long, 1)]
+    )
+    updated_parameters = {"b": 1}
+    updated_parameters.update(test_parameters)
+    assert _enforce_params_schema(test_parameters, test_schema) == updated_parameters
+
+    # Converting parameters keys to string if it is not
+    test_parameters = {1: 1.0}
+    test_schema = ParamSchema([ParamSpec("1", DataType.double, 1.0)])
+    assert _enforce_params_schema(test_parameters, test_schema) == {"1": 1.0}
+
+    # Raise error if parameters is not dictionary
+    with pytest.raises(MlflowException, match=r"Parameters must be a dictionary. Got type 'int'."):
+        _enforce_params_schema(100, test_schema)
+
+    # Raise error if invalid parameters are passed
+    test_parameters = {"a": True, "b": (1, 2), "c": b"test"}
```

Comment:
Can we include some additional test cases, e.g. passing a dictionary, passing a float instead of an int, passing a Python object?

Reply:
@serena-ruan Do we also have a test cases for the following?:

- Passing a parameter that is not defined in the signature throws exception
- Passing a parameter to an MLflow Model Server that is of the wrong type / that is not defined in the signature returns 4xx

If not, can we add them?

Reply:
Sure, but for case 1 `Passing a parameter that is not defined in the signature throws exception` we ignore the extra parameters and raise a warning instead of an exception.

Reply:
^ Sounds good!